include/h2o/pipe_utils.h

```diff
@@ -0,0 +1,39 @@
+#ifndef __H2O__PIPE_UTILS__H__
+#define __H2O__PIPE_UTILS__H__
+
+#include <stddef.h>
+#include <stdbool.h>
+
+enum h2o_send_state;
+struct st_h2o_req_t;
+struct st_h2o_context_t;
+
+typedef struct st_h2o_pipe_reader_t {
+    int fds[2]; /* fd[0] set to -1 unless used */
+    int inflight;
+
+    size_t body_bytes_read, body_bytes_sent;
+} h2o_pipe_reader_t;
+
+int h2o_empty_pipe(int fd);
+
+void h2o_pipe_reader_send(struct st_h2o_req_t *req, h2o_pipe_reader_t *pipe_reader, enum h2o_send_state send_state);
+
+void h2o_pipe_reader_init(h2o_pipe_reader_t *pipe_reader);
+
+int h2o_pipe_reader_new(struct st_h2o_context_t *ctx, h2o_pipe_reader_t *pipe_reader);
+
+int h2o_pipe_reader_start(struct st_h2o_context_t *ctx, h2o_pipe_reader_t *pipe_reader);
+
+void h2o_pipe_reader_dispose(struct st_h2o_context_t *ctx, h2o_pipe_reader_t *pipe_reader);
+
+int h2o_pipe_reader_is_empty(h2o_pipe_reader_t *pipe_reader);
+
+void h2o_pipe_reader_update(h2o_pipe_reader_t *pipe_reader, size_t read_body_bytes);
+
+static inline bool h2o_pipe_reader_is_initialized(h2o_pipe_reader_t *pipe_reader)
```

Comment:
I think the name is incorrect, as this function returns false when the file descriptor is set to -1, even though the "init" function initializes the value to -1.

Maybe something like "is_in_use"?